feat: add Futureswap volume and fees adapters

brenopolanski opened the pull request:
## Summary

- Closes: https://github.com/DefiLlama/dimension-adapters/issues/5144

Adds volume and fees tracking for Futureswap on Arbitrum and Avalanche.

- `dexs/futureswap.ts` - Tracks trading volume from `PositionChanged` events
- `fees/futureswap.ts` - Tracks trading fees (0.05% protocol fee) with revenue distribution
- Uses official contract addresses from [Futureswap documentation](https://docs.futureswap.com/protocol/developer/addresses-abis-and-links)

## Notes

- The adapter tracks only the Futureswap protocol fee (0.05%). The separate 0.05% Uniswap v3 fee paid to Uniswap LPs is not included (https://docs.futureswap.com/protocol/trading/fees).

llamabutler commented:
The futureswap.ts adapter exports:
        
 
 
    > adapters@1.0.0 test
    > ts-node --transpile-only cli/testAdapter.ts dexs futureswap.ts
    
    🦙 Running FUTURESWAP.TS adapter 🦙
    ---------------------------------------------------
    Start Date:	Tue, 16 Dec 2025 12:54:02 GMT
    End Date:	Wed, 17 Dec 2025 12:54:02 GMT
    ---------------------------------------------------
    
    chain     | Daily volume | Start Time
    ---       | ---          | ---       
    arbitrum  | 0.00         | 13/10/2021
    avax      | 0.00         | 22/4/2022 
    Aggregate | 0.00         |

llamabutler commented:
The futureswap.ts adapter exports:
        
 
 
    > adapters@1.0.0 test
    > ts-node --transpile-only cli/testAdapter.ts dexs futureswap.ts
    
    🦙 Running FUTURESWAP.TS adapter 🦙
    ---------------------------------------------------
    Start Date:	Tue, 16 Dec 2025 12:56:59 GMT
    End Date:	Wed, 17 Dec 2025 12:56:59 GMT
    ---------------------------------------------------
    
    chain     | Daily volume | Start Time
    ---       | ---          | ---       
    arbitrum  | 0.00         | 13/10/2021
    avax      | 0.00         | 22/4/2022 
    Aggregate | 0.00         |

noateden commented on fees/futureswap.ts:
```diff
@@ -0,0 +1,115 @@
+  }
+
+  const USDC =
+    chain === CHAIN.ARBITRUM
```
same here, should add to config

llamabutler commented:
The futureswap.ts adapter exports:
        
 
 
    > adapters@1.0.0 test
    > ts-node --transpile-only cli/testAdapter.ts fees futureswap.ts
    
    🦙 Running FUTURESWAP.TS adapter 🦙
    ---------------------------------------------------
    Start Date:	Wed, 17 Dec 2025 21:30:09 GMT
    End Date:	Thu, 18 Dec 2025 21:30:09 GMT
    ---------------------------------------------------
    
    chain     | Daily fees | Daily user fees | Daily revenue | Daily protocol revenue | Daily supply side revenue | Daily holders revenue | Start Time
    ---       | ---        | ---             | ---           | ---                    | ---                       | ---                   | ---       
    arbitrum  | 0.00       | 0.00            | 0.00          | 0.00                   | 0.00                      | 0.00                  | 13/10/2021
    avax      | 0.00       | 0.00            | 0.00          | 0.00                   | 0.00                      | 0.00                  | 22/4/2022 
    Aggregate | 0.00       | 0.00            | 0.00          | 0.00                   | 0.00                      | 0.00                  |

brenopolanski commented:
All done @noateden. Could you please review it again?

llamabutler commented:
The futureswap.ts adapter exports:
        
 
 
    > adapters@1.0.0 test
    > ts-node --transpile-only cli/testAdapter.ts dexs futureswap.ts
    
    🦙 Running FUTURESWAP.TS adapter 🦙
    ---------------------------------------------------
    Start Date:	Thu, 18 Dec 2025 12:21:36 GMT
    End Date:	Fri, 19 Dec 2025 12:21:36 GMT
    ---------------------------------------------------
    
    chain     | Daily volume | Daily fees | Daily revenue | Daily supply side revenue | Daily protocol revenue | Daily holders revenue | Start Time
    ---       | ---          | ---        | ---           | ---                       | ---                    | ---                   | ---       
    arbitrum  | 0.00         | 0.00       | 0.00          | 0.00                      | 0.00                   | 0.00                  | 13/10/2021
    avax      | 0.00         | 0.00       | 0.00          | 0.00                      | 0.00                   | 0.00                  | 22/4/2022 
    Aggregate | 0.00         | 0.00       | 0.00          | 0.00                      | 0.00                   | 0.00                  |

llamabutler commented:
The futureswap.ts adapter exports:
        
 
 
    > adapters@1.0.0 test
    > ts-node --transpile-only cli/testAdapter.ts dexs futureswap.ts
    
    🦙 Running FUTURESWAP.TS adapter 🦙
    ---------------------------------------------------
    Start Date:	Thu, 18 Dec 2025 12:22:03 GMT
    End Date:	Fri, 19 Dec 2025 12:22:03 GMT
    ---------------------------------------------------
    
    chain     | Daily volume | Daily fees | Daily revenue | Daily supply side revenue | Daily protocol revenue | Daily holders revenue | Start Time
    ---       | ---          | ---        | ---           | ---                       | ---                    | ---                   | ---       
    arbitrum  | 0.00         | 0.00       | 0.00          | 0.00                      | 0.00                   | 0.00                  | 13/10/2021
    avax      | 0.00         | 0.00       | 0.00          | 0.00                      | 0.00                   | 0.00                  | 22/4/2022 
    Aggregate | 0.00         | 0.00       | 0.00          | 0.00                      | 0.00                   | 0.00                  |

noateden reviewed (changes requested):
hi @brenopolanski, I did refactor a bit. Can you please verify numbers? I think we over-counting them for example

```
🦙 Running FUTURESWAP adapter 🦙
---------------------------------------------------
Start Date:	Wed, 10 Jan 2024 00:00:00 GMT
End Date:	Thu, 11 Jan 2024 00:00:00 GMT
---------------------------------------------------

chain     | Daily volume | Daily fees | Daily revenue | Daily supply side revenue | Daily protocol revenue | Daily holders revenue | Start Time
---       | ---          | ---        | ---           | ---                       | ---                    | ---                   | ---       
arbitrum  | 144.67 B     | 44.41 M    | 8.88 M        | 35.52 M                   | 8.88 M                 | 0.00                  | 13/10/2021
avax      | 0.00         | 0.00       | 0.00          | 0.00                      | 0.00                   | 0.00                  | 22/4/2022 
Aggregate | 144.67 B     | 44.41 M    | 8.88 M        | 35.52 M                   | 8.88 M                 | 0.00                  |           

```

brenopolanski commented:
Hey @noateden, I've updated the adapter accordingly.

- Removed volume entirely (`PositionChanged` doesn't map to spot volume)
- Fees are now tracked conservatively using only `tradeFee`
- Treating 100% of fees as protocol revenue
- Moved the adapter to the `fees/` folder

Tests look consistent for Arbitrum. Let me know if you would like another update.

```
pnpm test fees futureswap 2024-01-11

> adapters@1.0.0 test /Users/brenopolanski/github/dev/llamacorp/job/dimension-adapters
> ts-node --transpile-only cli/testAdapter.ts "fees" "futureswap" "2024-01-11"

🦙 Running FUTURESWAP adapter 🦙
---------------------------------------------------
Start Date:	Wed, 10 Jan 2024 00:00:00 GMT
End Date:	Thu, 11 Jan 2024 00:00:00 GMT
---------------------------------------------------

chain     | Daily fees | Daily revenue | Daily protocol revenue | Start Time
---       | ---        | ---           | ---                    | ---
arbitrum  | 44.41 M    | 44.41 M       | 44.41 M                | 13/10/2021
avax      | 0.00       | 0.00          | 0.00                   | 22/4/2022
Aggregate | 44.41 M    | 44.41 M       | 44.41 M                |
```

noateden commented on fees/futureswap.ts:
```diff
@@ -0,0 +1,79 @@
+    'All reported trade fees are treated as protocol revenue due to lack of on-chain fee distribution data.',
+  ProtocolRevenue:
+    'Same as Revenue.',
+  Limitations:
```
what are you trying to explain here? can you remove?
